47 send email including interview info to jobseeker

SMSaeedi opened the pull request.

SMSaeedi committed:
send-email-including-interview-info-to-jobseeker

SMSaeedi committed:
send-email-including-interview-info-to-jobseeker
Separate email service from interview service in terms of agenda
Rename email template, dedicated to job-seeker
Added time slot availability to avoid being booked more than one time
Unit tests coverages

SMSaeedi committed:
send-email-including-interview-info-to-jobseeker

m-tilab commented on src/main/resources/application.yml:
```diff
@@ -17,6 +17,19 @@ spring:
         format_sql: true
+
+  mail:
+    host: smtp.linchpino.com
+    port: 587
+    username: no-reply@linchpino.com
+    password: NoreplyLinchP!no
```
Please assign the host, username, and password as below, as I mentioned twice before:
host: ${MAIL_HOST:smtp.linchpino.com}

m-tilab commented on src/main/resources/application.yml:
```diff
@@ -46,3 +59,5 @@ linkedin:
   policyRegex: "^(?=.*[A-Za-z])(?=.*\\d)(?=.*[@$!%*?&])[A-Za-z\\d@$!%*?&]{6,}$"
 
+application:
+  url: "https://linchpino.liara.run"
```
Also for application.url property

m-tilab commented on src/main/resources/mailTemplate/email-template.html:
```diff
@@ -0,0 +1,28 @@
+<!DOCTYPE html>
+<html lang="en" xmlns:th="http://www.w3.org/1999/xhtml">
+<head>
+    <meta charset="UTF-8">
+    <meta name="viewport" content="width=device-width, initial-scale=1.0">
+    <title>Confirmation of Interview Schedule on Linchpino</title>
+</head>
+<body>
```
The mail template doesn't need HTML, head and body because it does show in  the browser individually

m-tilab commented on src/main/kotlin/com/linchpino/core/service/EmailService.kt:
```diff
@@ -0,0 +1,29 @@
+        val message = emailSender.createMimeMessage()
+        val helper = MimeMessageHelper(message, true, "UTF-8")
+        helper.setTo(to)
+//        helper.setSubject("Confirmation of Interview Schedule on Linchpino")
```
Where we set subject to the mail?

SMSaeedi replied:
Into the title tag in template file

m-tilab reviewed (changes requested):
Please merge the development branch to your feature branch. Ty

SMSaeedi committed:
fix - after review and comments

SMSaeedi committed:
unit test fixed

SMSaeedi committed:
unit test fixed
removed unnecessary jar file from pom.xml

SMSaeedi committed:
create interview test, bug fix, send email to job seeker test

SMSaeedi committed:
sendEmail test

Commit:
47: Add app password authentication for gmail and some minor changes on MailService

Commit:
47: Add app password authentication for gmail and some minor changes on MailService

SMSaeedi committed:
unit tests

SMSaeedi committed:
unit tests

SMSaeedi committed:
Separate email service from interview service in terms of agenda
Rename email template, dedicated to job-seeker
Added time slot availability to avoid being booked more than one time
Unit tests coverages

SMSaeedi committed:
small changes, happened during rebasing

SMSaeedi committed:
changes after rebase with develop

m-tilab commented on src/main/kotlin/com/linchpino/core/dto/InterviewDTO.kt:
```diff
@@ -9,7 +9,7 @@ data class CreateInterviewRequest(
     @field:NotNull(message = "interviewTypeId is required") val interviewTypeId: Long,
     @field:NotNull(message = "timeSlotId is required") val timeSlotId: Long,
-    @field:NotNull(message = "mentorAccId is required") val mentorAccId: Long,
+    @field:NotNull(message = "mentorAccId is required") val mentorAccountId: Long,
```
Did you inform @HosseinRA1 about this change?

SMSaeedi replied:
@HosseinRA1 Hi, I have changed the name of **mentorAccId** to **mentorAccountId** in order to be more human readable, kindly consider this in your frontend side implementation.

m-tilab commented on src/main/kotlin/com/linchpino/core/service/EmailService.kt:
```diff
@@ -0,0 +1,72 @@
+    @Value("\${application.url}")
+    var applicationUrl: String? = null
+
+    fun sendEmail(interview: Interview) {
```
Please use a more descriptive method name because this method is only used for sending an interview mail to the job seeker

SMSaeedi replied:
Done!

m-tilab commented on src/main/kotlin/com/linchpino/core/service/EmailService.kt:
```diff
@@ -0,0 +1,72 @@
+        )
+    }
+
+    fun sendTemplateEmail(
```
Can we make this method as private?

SMSaeedi replied:
made it internal, couldn't be private due to the UT.

m-tilab commented on src/main/resources/application.yml:
```diff
@@ -17,6 +17,21 @@ spring:
+    host: ${MAIL_SERVER_URL:smtp.gmail.com}
+    port: 587
+    username: ${MAIL_USERNAME:no-reply@linchpino.com}
+    password: ${MAIL_PASSWORD:cwwwvdugodzgnily}
```
I defined "MAIL_PASSWORD" as a GitHub secret. Please replace password value with "Password" to protect it

SMSaeedi replied:
Authentication Failed!
![photo_5960552783461597863_y](https://github.com/linchpino/linchpino-core/assets/45897831/3a3539ad-e7c4-4b59-beb6-ace88b90bd8d)

m-tilab commented on src/main/resources/application.yml:
```diff
@@ -46,3 +61,5 @@ linkedin:
 password:
   policyRegex: "^(?=.*[A-Za-z])(?=.*\\d)(?=.*[@$!%*?&])[A-Za-z\\d@$!%*?&]{6,}$"
 
+application:
+  url:  ${APPLICATION_URL:https://linchpino.liara.run}
```
It can be inline instead of two lines: "application.url:"

SMSaeedi replied:
it's not possible, @Value cannot detect it, when it's inline. A line above is also one element password policyRegex.

m-tilab commented on src/main/kotlin/com/linchpino/core/service/InterviewService.kt:
```diff
@@ -32,30 +34,42 @@ class InterviewService(
+            )
+
         val interview = populateInterviewObject(request, jobSeekerAccount)
         interviewRepository.save(interview)
```
Please update the Timeslot to status MentorTimeSlotEnum.ALLOCATED after save the repository and make sure the method is transactional

SMSaeedi replied:
Done!

m-tilab commented on src/main/kotlin/com/linchpino/core/service/InterviewService.kt:
```diff
@@ -32,30 +34,42 @@ class InterviewService(
+                        lastName = "",
+                        email = request.jobSeekerEmail,
+                        password = null,
+                        status = AccountStatusEnum.DEACTIVATED,
+                        type = AccountTypeEnum.JOB_SEEKER.value
```
the status and type fields are misordered as mentioned in the accountService.createAccount method

m-tilab replied:
I have talked to @sharifyy about it. Here Mohammad used the named argument feature of Kotlin and its correct without change

m-tilab reviewed (changes requested):
I have tested 3 scenarios:

1. a new job seeker creates an interview: failed
![image](https://github.com/linchpino/linchpino-core/assets/6872871/affb455d-3360-4cdd-8292-057a08ab782d)

![image](https://github.com/linchpino/linchpino-core/assets/6872871/b289aeaa-957e-4b82-9b2d-c15d8e48bdf9)
Two points need to be correct: 

- first, if the first and the lastname are empty, use: Dear "Jobseeker"
 ....
 - second: The condition activation link does not appear 

2. an old job seeker creates an interview: Approved
3. a new job seeker creates an interview that the timeslot has already taken: Approved

SMSaeedi committed:
PR comments

SMSaeedi committed:
repoName had been changed, rebase effects

SMSaeedi committed:
rebase effects

SMSaeedi committed:
PR comment

Commit:
47: fixed jobseeker-activation link and some small other fixes

sharifyy committed:
110 get all interview types api mentor registration  (#57)

* Search InterviewTypes.

* 124: improve SonarCloud Github Action

* 124: improve SonarCloud Github Action

* 124: improve SonarCloud Github Action

* 124: improve SonarCloud Github Action

---------

Co-authored-by: mehdi.tilab <mehdi.tilab@virtudev.net>

Commit:
Merge branch '47-send-email-including-interview-info-to-jobseeker' of github.com:linchpino/linchpino-core into 47-send-email-including-interview-info-to-jobseeker

Commit:
47: fix integration test

sharifyy commented on src/main/kotlin/com/linchpino/core/config/ThymeleafConfig.kt:
```diff
@@ -0,0 +1,29 @@
+import org.thymeleaf.templateresolver.ClassLoaderTemplateResolver
+import org.thymeleaf.templateresolver.ITemplateResolver
+
+@Configuration
```
Doesn't spring boot auto configure Thymeleaf?

m-tilab replied:
https://stackoverflow.com/questions/54789203/what-is-the-difference-between-springresourcetemplateresolver-and-classloadertem#:~:text=5-,You%20should%20be%20using%20SpringResourceTemplateResolver%20for%20web%20projects%20(using%20SpringMvc%20controllers)%2C%20and%20ClassLoaderTemplateResolver%20for%20command%20line%20projects%2C%20or%20when%20you%20need%20to%20generate%20html%20for%20non-web%20applications%20(such%20as%20sending%20html%20email%20or%20something%20like%20that).,-First%20error%20occured

By default Spring boot configured resolver by SpringResourceTemplateResolver instead of ClassLoaderTemplateResolver

sharifyy commented on src/main/kotlin/com/linchpino/core/dto/InterviewDTO.kt:
```diff
@@ -9,7 +9,7 @@ data class CreateInterviewRequest(
     @field:NotNull(message = "jobPositionId is required") val jobPositionId: Long,
     @field:NotNull(message = "interviewTypeId is required") val interviewTypeId: Long,
     @field:NotNull(message = "timeSlotId is required") val timeSlotId: Long,
-    @field:NotNull(message = "mentorAccId is required") val mentorAccId: Long,
```
Also change mentorAccId in message.

m-tilab replied:
Done

sharifyy commented on src/main/kotlin/com/linchpino/core/repository/InterviewRepository.kt:
```diff
@@ -11,6 +11,12 @@ import org.springframework.stereotype.Repository
 
 @Repository
 interface InterviewRepository : JpaRepository<Interview, Long> {
+    @Query(
```
This query does not need Case and you can write more simple query for it. Also use multiline string.

sharifyy commented on src/main/kotlin/com/linchpino/core/service/AccountService.kt:
```diff
@@ -26,6 +26,7 @@ import org.springframework.stereotype.Service
 import org.springframework.transaction.annotation.Transactional
 import java.time.ZoneOffset
 import java.time.ZonedDateTime
+import java.util.*
```
Remove * import.

m-tilab replied:
Done

sharifyy commented on src/main/kotlin/com/linchpino/core/service/EmailService.kt:
```diff
@@ -0,0 +1,82 @@
+    var applicationUrl: String? = null
+
+    fun sendingInterviewInvitationEmailToJobSeeker(interview: Interview) {
+        val fullName = "${interview.jobSeekerAccount!!.firstName} ${interview.jobSeekerAccount!!.lastName}"
```
jobSeekerAccount!!.firstName may throw null pointer.

sharifyy replied:
Use !! in production code with more caution. These objects are all nullable and may throw null pointer exception. Unless you are sure that these objects never take null value.

sharifyy commented on src/main/kotlin/com/linchpino/core/service/EmailService.kt:
```diff
@@ -0,0 +1,82 @@
+        val time = interview.timeSlot!!.fromTime.toLocalTime()
+        val zone = interview.timeSlot!!.fromTime.getZone()
+        val isJobSeekerIsNotActivated = interview.jobSeekerAccount!!.status != AccountStatusEnum.ACTIVATED
+        val jobSeekerExternalId = interview.jobSeekerAccount!!.externalId.toString()
```
externalId is of type String, therefore toString is not needed.

sharifyy commented on src/main/kotlin/com/linchpino/core/service/EmailService.kt:
```diff
@@ -0,0 +1,82 @@
+
+        val mailFrom: String? = environment.getProperty("spring.mail.properties.mail.smtp.from")
+        val mailFromName: String = environment.getProperty("mail.from.name", "Linchpino")
+
```
Use @Value annotation instead of calling getProperty on environment.

m-tilab replied:
Done

sharifyy commented on src/main/kotlin/com/linchpino/core/service/EmailService.kt:
```diff
@@ -0,0 +1,82 @@
+    }
+
+    internal fun generateTemplate(model: Map<String, Any>, templateName: String): String {
+
```
Why do you need internal access modifier here?

sharifyy commented on src/main/kotlin/com/linchpino/core/service/InterviewService.kt:
```diff
@@ -30,30 +33,49 @@ class InterviewService(
+            ?: accountRepository.findReferenceById(
+                accountService.createAccount(
+                    CreateAccountRequest(
+                        firstName = "",
```
Why empty string?

sharifyy commented on src/main/kotlin/com/linchpino/core/service/InterviewService.kt:
```diff
@@ -30,30 +33,49 @@ class InterviewService(
-        val mentorAcc = accountRepository.findReferenceById(createInterviewRequest.mentorAccId)
+        val mentorAcc = accountRepository.findReferenceById(createInterviewRequest.mentorAccountId)
         val typeInterview = interviewTypeRepository.findReferenceById(createInterviewRequest.interviewTypeId)
+        val isTimeSlotBooked = interviewRepository.isTimeSlotBooked(createInterviewRequest.timeSlotId)
```
You don't need two queries to check time-slot status and fetch time-slot. After fetching time-slot, you can just check if the time-slot is available or not

m-tilab replied:
Done!

sharifyy commented on src/main/kotlin/com/linchpino/core/service/TimeSlotService.kt:
```diff
@@ -27,4 +33,10 @@ class TimeSlotService(private val accountRepository: AccountRepository, private
         repository.saveAll(mentorTimeSlots)
     }
+
+    fun updateTimeSlotStatus(timeSlot: MentorTimeSlot, allocated: MentorTimeSlotEnum) {
```
Rename second parameter name.

m-tilab replied:
Done

sharifyy commented on src/main/kotlin/com/linchpino/core/service/EmailService.kt:
```diff
@@ -0,0 +1,82 @@
+
+    fun sendingInterviewInvitationEmailToJobSeeker(interview: Interview) {
+        val fullName = "${interview.jobSeekerAccount!!.firstName} ${interview.jobSeekerAccount!!.lastName}"
+        val finalName = fullName.takeIf { it.isEmpty() } ?: "Jobseeker"
```
finalName is always "Jobseeker"

sharifyy commented on src/test/kotlin/com/linchpino/core/service/InterviewServiceTest.kt:
```diff
@@ -47,36 +47,34 @@ class InterviewServiceTest {
-        // Given
         val jobSeekerAccount = Account().apply {
-            id = 2
+            id = 1
```
Having two accounts(mentor and job-seeker) with the same ID is semantically incorrect, although you are not persisting them in the database.

sharifyy commented on src/test/kotlin/com/linchpino/core/service/InterviewServiceTest.kt:
```diff
@@ -141,52 +146,57 @@ class InterviewServiceTest {
-            id = 1
+            id = 2
             name = "Test Interview Type"
         }
```
I don't get the intention behind changing all these IDs. Could you please elaborate?

Commit:
47: review fixes

Commit:
47: update email data

Commit:
47: reiview changes

Commit:
Merge branch '47-send-email-including-interview-info-to-jobseeker' of github.com:linchpino/linchpino-core into 47-send-email-including-interview-info-to-jobseeker

Commit:
Fix interview service test.

Commit:
Fix interview controller IT tests.

Commit:
Refactor InterviewServiceTest to include email service and timeSlotService interactions

Commit:
EmailService unit tests.

Commit:
47: Update docker-compose to add mail_password ENV

Commit:
47: Update docker-compose to add mail_password ENV

Commit:
47: fix property names
